fix: convert blob.subscription into Hamt

joewagner opened the pull request:
## Overview
The `Blob` stuct was an ipld hamt that contained a HashMap of its `Subcribers`, which contained a HashMap of `SubscriptionGroups`.  These nested hash maps had the potential to grow larger than can be flushed to the blockstore (1MB).
This change converts the nested hash maps into nested hamts, hence preventing issues with exceeding the 1MB limit

## Details 
This change include a lot of conversion of hash map to hamt while maintaining logic.  The majority of this is working through when to flush the inner hamts and converting from `Option` to `Result`.  It also includes some additions to the hamt implementation.  A hamt now has an `iter` method so it can be used in for loops, and the hamt can now be keyed with a `String`, which is what the subscription groups are keyed with. 

fixes #441

joewagner commented on fendermint/actors/blobs/shared/src/state.rs:
```diff
@@ -279,25 +318,64 @@ impl fmt::Display for SubscriptionId {
+}
+
+// TODO: is works well afaict, but is it a ok/good/reasonable pattern in Rust?
+fn sub_illegal_state_err_map<E>(e: E, sub_id: &SubscriptionId) -> ActorError
```
I did try this snippet (and few other things), but `val` is a Result where the error is a hamt error, and `from_utf8` returns a Result where the error is a utf8 error.  As far as I can tell there isn't a way to map both of these errors to `ActorError` without two calls to `.map_err`.

avichalp replied:
ah alright

brunocalza replied:
the only thing here is that the errors are not really duplicates. For the first one the message "subscriptions for id {} cannot be iterated over: {}" makes sense. But for `from_utf8`, it does not. I don't really mind to have two `map_err` with the correct error message. The same applies for [`err_map`](https://github.com/recallnet/ipc/pull/520/files#diff-8f76f034a5bc328443f73e9c6bbb97e66dbaa18ef554803e85b4c4529b7d36a4R1425-R1434)

joewagner commented on recall/ipld/src/hamt/core.rs:
```diff
@@ -284,6 +290,22 @@ impl MapKey for Vec<u8> {
     }
 }
 
+impl MapKey for String {
```
I did consider this briefly.  What I found was that the `fendermint_actors_blobs_shared` package has `recall_ipld` as a dependency, so making the `SubscriptionId` type available to this impl would create a circular dependency.
Do we need to move any types we want to use as hamt keys into a separate package? 
FYI: the same sort of issue exists for #538 with the `BlobSourceSet` being used as a key.

joewagner commented on fendermint/actors/blobs/shared/src/state.rs:
```diff
@@ -207,18 +210,54 @@ impl From<u64> for Hash {
+    size: u64,
+}
+
+impl BlobSubscribers {
```
@ukstv I'm not very experienced with Cargo, so I'm probably missing something super simple.  Or maybe I'm not understanding your suggestion, but since `fendermint_actors_blobs_shared` is a package nested inside of the `fendermint_actors_blobs` package, I don't know how to import `BlobSubscribers` into this file so it can be used by `Blob` on line 221.
Are you suggesting I should move this struct and impl out of the `shared` directory?

avichalp replied:
oh sorry for the confusion but since `BlobSubscribers` is used in the the shared state itself doesn't it makes more sense to leave this here? other HAMTs are defined in the top level state but maybe it is fine to define this one here @ukstv?

brunocalza replied:
Yeah, I'm facing a similar situation on trying to hamt-ify approvals. We can't actually move it to `blobs/src/state/blobs.rs` because we would face a circular dependency, because `blobs/src/state/blobs.rs` already depends on `blobs/shared/src/state.rs`. I think it's fine to leave it here.

joewagner replied:
Ok, that sounds good to me.  I experimented moving it, and the only thing I could think of that might work was a third package which does not seem needed.

joewagner commented:
I rebased onto `main` and fixed the hamt name issue.  The other two requests result in a circular dependency issue when I try to fix.  @ukstv can you take another look?

brunocalza commented on fendermint/actors/blobs/src/state.rs:
```diff
@@ -1010,20 +1040,21 @@ impl State {
-                )))?;
+        let mut subscribers = blob.subscribers.hamt(store)?;
+        let mut group = subscribers
+            .get(&subscriber)
```
maybe you missed a question mark here? instead of `unwrap`

joewagner replied:
Good call, this is fixed.

brunocalza commented on fendermint/actors/blobs/src/state.rs:
```diff
@@ -900,13 +933,10 @@ impl State {
-                        .subscriptions
-                        .get(&id.to_string())
-                    {
+                    // TODO: I'm not sure about using all these unwraps. It seems like this function should return a Result<Option...>
```
yeah, I think the ideal would be to return `Result` and add `?`

brunocalza reviewed (approved):
This looks good @joewagner  I made 2 final comments:
- my opinion on the error mapping
- adding `Result` to the return of `get_blob_status `

joewagner commented:
> This looks good @joewagner I made 2 final comments:
> 
> * my opinion on the error mapping
> * adding `Result` to the return of `get_blob_status `

Thanks for looking again.  Both things sound good. I'll make these changes now.

sanderpick commented:
@ukstv can you give this another once over please?

ukstv reviewed (approved):
@sanderpick Approved!

sanderpick commented:
@joewagner looks like docker publish failed... can you take a look?

joewagner commented:
> @joewagner looks like docker publish failed... can you take a look?

for sure

joewagner commented:
> @joewagner looks like docker publish failed... can you take a look?

@sanderpick The original failure was during a call to 
```
apt-get install -y g++-aarch64-linux-gnu libc6-dev-arm64-cross; \
  rustup target add aarch64-unknown-linux-gnu; \
  rustup toolchain install ${RUST_VERSION}-aarch64-unknown-linux-gnu; \
  fi
```
 in the [Dockerfile](https://github.com/recallnet/ipc/blob/main/fendermint/docker/builder.ci.Dockerfile#L63).  The only message indicated that the failure was caused by `Error 1`.  

I couldn't figure out why the `apt-get` command would fail, so I re-ran the job. The second run's error message is `System.IO.IOException: No space left on device : '/home/ghrunner/actions-runner/_diag/Worker_20250303-175315-utc.log' Un`
Is the runner out of disk space?

sanderpick commented:
hmm, doesn't look like it. here's runner 1:

```
ghrunner@c2-xm-f37dd4:~$ df -h
Filesystem             Size  Used Avail Use% Mounted on
tmpfs                  101G   32M  101G   1% /run
/dev/nvme0n1p3         439G  309G  108G  75% /
tmpfs                  502G     0  502G   0% /dev/shm
tmpfs                  5.0M     0  5.0M   0% /run/lock
/dev/nvme0n1p2         511M  6.1M  505M   2% /boot/efi
tmpfs                  101G  4.0K  101G   1% /run/user/0
tmpfs                  101G  4.0K  101G   1% /run/user/1007
pool1                  512T  128K  512T   1% /pool1
pool1/testnet1-blocks  1.0T   42G  983G   5% /mnt/testnet1/blocks
pool1/testnet1-data     50T  9.2G   50T   1% /mnt/testnet1/data
```

and runner 2:
```
ghrunner@c2-xm-f37e35:~$ df -h
Filesystem             Size  Used Avail Use% Mounted on
tmpfs                  101G   34M  101G   1% /run
/dev/nvme0n1p3         439G  290G  127G  70% /
tmpfs                  502G     0  502G   0% /dev/shm
tmpfs                  5.0M     0  5.0M   0% /run/lock
/dev/nvme0n1p2         511M  6.1M  505M   2% /boot/efi
tmpfs                  101G  4.0K  101G   1% /run/user/0
tmpfs                  1.0M     0  1.0M   0% /var/snap/lxd/common/ns
pool1                  512T  128K  512T   1% /pool1
pool1/testnet1-blocks  1.0T   47G  978G   5% /mnt/testnet1/blocks
pool1/testnet1-data     50T  164M   50T   1% /mnt/testnet1/data
tmpfs                  101G  4.0K  101G   1% /run/user/1009
```

i'll take a look tomorrow

sanderpick commented:
or maybe @sam701 can... he'll need the image for the testnet deploy

sanderpick commented:
i tried `docker prune`ing on both boxes... def clean a lot of space. i re-kicked the job, we'll see...
